lib/invoker/power/balancer.rb

```diff
@@ -43,29 +43,47 @@ def initialize(connection, protocol)
         @connection = connection
         @protocol = protocol
         @http_parser = HttpParser.new(protocol)
+        @response_http_parser = HttpParser.new(protocol)
         @session = nil
         @buffer = []
       end
 
       def install_callbacks
+        http_parser.on_url { |url| url_received(url) }
         http_parser.on_headers_complete { |headers| headers_received(headers) }
         http_parser.on_message_complete { |full_message| complete_message_received(full_message) }
+
+        @response_http_parser.on_message_complete { response_complete }
+
         connection.on_data { |data| upstream_data(data) }
         connection.on_response { |backend, data| backend_data(backend, data) }
         connection.on_finish { |backend, name| frontend_disconnect(backend, name) }
       end
 
+      def response_complete
+        # disconnect from backend after each response, since subsequent
+        # requests from persistent HTTP connections might be proxied to a stale
+        # backend
+        EventMachine.next_tick do # still needs chance to send data
+          connection.unbind_backend(@session)
+          @session = nil
+          @response_http_parser.reset
+          http_parser.reset
+        end
+      end
```

Comment:
This is sort of unrelated to this change isn't it? Do you think this should be a good idea in general or the path matching specifically made something worse, thereby highlighting a bug we haven't seen before?